### PR TITLE
Add DeeBERT (entropy-based early exiting for *BERT)

### PR DESCRIPTION
**Pull request by ji-xin**
Add DeeBERT (entropy-based early exiting for *BERT).
Paper: https://www.aclweb.org/anthology/2020.acl-main.204/
Based on its original repository: https://github.com/castorini/DeeBERT

### CODE REVIEW

File: examples/deebert/run_glue_deebert.py
Comment (JetRunner) [inline examples/deebert/run_glue_deebert.py:17]: Are `DeeBertForSequenceClassification` and `DeeRobertaForSequenceClassification` better?

File: examples/deebert/run_glue_deebert.py
Comment (JetRunner) [inline examples/deebert/run_glue_deebert.py:619]:
```suggestion
    elif args.model_type == "roberta":
```

File: examples/deebert/run_glue_deebert.py
Comment (JetRunner) [inline examples/deebert/run_glue_deebert.py:622]:
```suggestion
      else:
          raise NotImplementedError()
```

File: examples/deebert/run_glue_deebert.py
Comment (JetRunner) [inline examples/deebert/run_glue_deebert.py:683]: same here

File: examples/deebert/README.md
Comment (JetRunner) [inline examples/deebert/README.md:16]: See the comment for `requirements.txt`

File: examples/deebert/deebert/modeling_highway_bert.py
Comment (JetRunner) [inline examples/deebert/deebert/modeling_highway_bert.py:198]: https://github.com/huggingface/transformers/blob/97355339f69e75df75fa91670d79a1edc0060389/examples/bert-loses-patience/pabee/modeling_pabee_bert.py#L155

we now have a util function. you can refactor these codes

File: examples/deebert/deebert/modeling_highway_bert.py
Comment (JetRunner) [inline examples/deebert/deebert/modeling_highway_bert.py:238]: https://github.com/huggingface/transformers/blob/97355339f69e75df75fa91670d79a1edc0060389/examples/bert-loses-patience/pabee/modeling_pabee_bert.py#L173

we now have a util function. you can refactor these codes

File: examples/deebert/README.md
Comment (JetRunner) [inline examples/deebert/README.md:42]: You may want to elaborate a little on what variable is what in these scripts?

**Review by JetRunner — CHANGES_REQUESTED** [review]
Thanks for your contribution!

some high-level comments here:

1. how many iterations do your test run? You may want to reduce the `max_iteration` a little to make it faster.
2. `/examples/deebert/src/` seems to be better than `/examples/deebert/deebert/`

File: examples/deebert/requirements.txt
Comment (stefan-it) [reply examples/deebert/requirements.txt:3]: I think all of these are already specified in the top-level requirements from `examples` folder:

https://github.com/huggingface/transformers/blob/master/examples/requirements.txt

So there's no need to add it here (exept `boto3`)

**Comment by stefan-it** [comment]
Btw: would be awesome so see a token classification example 😅

File: examples/deebert/requirements.txt
Comment (JetRunner) [reply examples/deebert/requirements.txt:3]: Yes! @stefan-it w.r.t. https://github.com/huggingface/transformers/pull/5477#discussion_r449543899

**Comment by Ji-Xin** [comment]
Hi @JetRunner, thanks for the review! I have updated according to your suggestions.

**Comment by Ji-Xin** [comment]
2 checks fail, however they don't seem relevant to my commits.

**Review by JetRunner — APPROVED** [review]
Thanks @Ji-Xin ! It looks much better now!

Please wait for the final approval of @LysandreJik

File: examples/deebert/src/modeling_highway_bert.py
Comment (LysandreJik) [inline examples/deebert/src/modeling_highway_bert.py:27]: (nit) we don't need to specify that since we're not python 2 compatible
```suggestion
        super().__init__()
```

File: examples/deebert/src/modeling_highway_roberta.py
Comment (LysandreJik) [inline examples/deebert/src/modeling_highway_roberta.py:31]: Those should already be defined in the `DeeBertModel`?

**Review by LysandreJik — APPROVED** [review]
Cool, very clean!

**Comment by Ji-Xin** [comment]
@LysandreJik Thanks for the comments and I've updated accordingly!

**Pull request merged**
